gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -415,17 +397,20 @@ public void run() {
   }
 
   /**
-   * Add the given {@link WorkUnit} for execution.
+   * Prepare the {@link WorkUnit}s for execution by populating the job and task IDs.
    */
-  private void addWorkUnit(WorkUnit workUnit, int sequence, JobState jobState) {
-    workUnit.setProp(ConfigurationKeys.JOB_ID_KEY, this.jobContext.getJobId());
-    String taskId = JobLauncherUtils.newTaskId(this.jobContext.getJobId(), sequence);
-    workUnit.setId(taskId);
-    workUnit.setProp(ConfigurationKeys.TASK_ID_KEY, taskId);
-    jobState.addTask();
-    // Pre-add a task state so if the task fails and no task state is written out,
-    // there is still task state for the task when job/task states are persisted.
-    jobState.addTaskState(new TaskState(new WorkUnitState(workUnit)));
+  private void prepareWorkUnits(List<WorkUnit> workUnits, JobState jobState) {
+    int taskIdSequence = 0;
+    for (WorkUnit workUnit : workUnits) {
+      workUnit.setProp(ConfigurationKeys.JOB_ID_KEY, this.jobContext.getJobId());
+      String taskId = JobLauncherUtils.newTaskId(this.jobContext.getJobId(), taskIdSequence++);
```

Comment:
Previously MultiWorkUnits were handled differently. That seems to have changed now. Why? Wouldn't this create problems with existing job states?